caas/kubernetes/clientconfig/k8s.go

```diff
@@ -162,3 +161,60 @@ func getKubeConfigPath() string {
 	}
 	return envPath
 }
+
+// getKubeConfig reads kubeconfig file either stdin or config file, then deserializes the contents into Config object.
+func getKubeConfig() (*clientcmdapi.Config, error) {
+	var filename string
+	var raw []byte
+
+	info, err := os.Stdin.Stat()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if info.Mode()&os.ModeNamedPipe != 0 || info.Size() > 0 {
+		// Try to read from Stdin.
+		raw, err = ioutil.ReadAll(os.Stdin)
+	} else {
+		// Try to read from kubeconfig file because no Stdin.
+		filename = getKubeConfigPath()
+		raw, err = ioutil.ReadFile(filename)
+		if err != nil {
+			err = errors.Annotatef(err, "failed to read kubernetes config from '%s'", filename)
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := clientcmd.Load(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set LocationOfOrigin on every Cluster, User, and Context.
+	for key, obj := range config.AuthInfos {
+		obj.LocationOfOrigin = filename
+		config.AuthInfos[key] = obj
+	}
+	for key, obj := range config.Clusters {
+		obj.LocationOfOrigin = filename
+		config.Clusters[key] = obj
+	}
+	for key, obj := range config.Contexts {
+		obj.LocationOfOrigin = filename
+		config.Contexts[key] = obj
+	}
+
+	if config.AuthInfos == nil {
+		config.AuthInfos = map[string]*clientcmdapi.AuthInfo{}
```

Comment:
I am not sure about these? Don' you want to initialise CLuster, User and Context before setting of location?

And if not (say, for the cases where there are no clusters, users, contexts), what is the point of initializing them here? If it is to ensure that the returned config is valid, surely it needs to take place before populating the values?